### PR TITLE
run test-suite (and repack netlib) for blis

### PR DESCRIPTION
**Pull request by h-vetinari**

### CODE REVIEW

**Comment by h-vetinari** [comment]
Getting an error
```
  File "/opt/conda/lib/python3.8/site-packages/conda_build/render.py", line 475, in _simplify_to_exact_constraints
    raise ValueError("Conflicting exact pins: {}".format(exact_pins))
ValueError: Conflicting exact pins: [['3.9.0', '3_h92ddd45_netlib'], ['3.9.0', '3_h92ddd45_netlib'], ['3.9.0', '6_blis']]
```

Based on the number of resolution steps before, I'm guessing this happens for `blas-devel`. Let's see if https://github.com/conda-forge/blas-feedstock/pull/62/commits/d80fe2b7282af13db0f6469e9bf1748d45d22744 helps.

**Comment by isuruf** [comment]
FWIW, I don't like repackaging the netlib variant of lapack as blis. With the current scheme, it's obvious that we are using BLAS from blis and LAPACK from netlib. With the new scheme, it's not obvious whether LAPACK comes from netlib or flame or other source.

**Comment by h-vetinari** [comment]
> FWIW, I don't like repackaging the netlib variant of lapack as blis. With the current scheme, it's obvious that we are using BLAS from blis and LAPACK from netlib. With the new scheme, it's not obvious whether LAPACK comes from netlib or flame or other source.

It's a point I expected to discuss with you. ;-)

But right now I was just trying to figure out how to run the test suite for blis here... Not sure if conda could be taught to correctly resolve the exact subpackage pins mixed with other variants, but so far, the only thing I've gotten working is a repack.

I had an intermediate version (lost since the rebase) of renaming the `lapack(e)`-outputs for `blis` only, e.g. it could be something like `lapack(e)-3.9.0-7_netlib_repacked_for_blis`, which would still work with the `*blis` selector. WDYT?

**Comment by h-vetinari** [comment]
Also, I now have a running test suite for blis on windows, and the results are _bad_: `13% tests passed, 90 tests failed out of 103`. This probably explains why I'm having [~440 errors](https://github.com/conda-forge/scipy-feedstock/pull/153) in the scipy test suite for blis on windows.

**Comment by h-vetinari** [comment]
> @h-vetinari: Also, I now have a running test suite for blis on windows, and the results are _bad_: `13% tests passed, 90 tests failed out of 103`. This probably explains why I'm having [~440 errors](https://github.com/conda-forge/scipy-feedstock/pull/153) in the scipy test suite for blis on windows.

On second thought - all the `BLAS*` tests pass, all the `LAPACK*` tests fail - turns out the integration isn't working yet after all...

File: recipe/meta.yaml [hunk 1]
```diff
@@ -192,13 +203,7 @@ outputs:
...
         - openblas   # [blas_impl == "openblas"]
         - mkl-devel  # [blas_impl == "mkl"]
-        - {{ pin_subpackage("liblapack", exact=True) }}      # [blas_impl != 'blis']
-        - {{ pin_subpackage("liblapacke", exact=True) }}     # [blas_impl != 'blis']
-        - liblapack  {{ version }} *netlib                   # [blas_impl == 'blis']
-        - liblapacke {{ version }} *netlib                   # [blas_impl == 'blis']
-        - {{ pin_subpackage("libcblas", exact=True) }}
-        - {{ pin_subpackage("libblas", exact=True) }}
-        - {{ pin_subpackage("blas-devel", exact=True) }}
+        - {{ pin_subpackage("blas", exact=True) }}
```
Comment (h-vetinari) [inline recipe/meta.yaml:206]: @isuruf, not sure I understood your intention from #61 here correctly, but `blas-devel` depending on itself sounds fragile and circular. My understanding would be that it builds on `blas` and adds the "development headers" (or whatever is the equivalent)...?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -226,23 +231,15 @@ outputs:
...
-        - blis 0.8.0
-      {% endif %}
       run:
-        - {{ pin_subpackage("liblapack", exact=True) }}      # [blas_impl != 'blis']
-        - {{ pin_subpackage("liblapacke", exact=True) }}     # [blas_impl != 'blis']
-        - liblapack  {{ version }} *netlib                   # [blas_impl == 'blis']
-        - liblapacke {{ version }} *netlib                   # [blas_impl == 'blis']
-        - {{ pin_subpackage("libcblas", exact=True) }}
         - {{ pin_subpackage("libblas", exact=True) }}
-        - {{ pin_subpackage("blas-devel", exact=True) }}
```
Comment (h-vetinari) [inline recipe/meta.yaml:245]: Same here, not sure why `blas` should need e.g. `mkl-devel`?

**Review by h-vetinari — COMMENTED** [review]
@isuruf, this should be ready for a first round of discussions - not saying repacking is a must, but I believe (with the build string distinction) it's definitely feasible without ambiguity.

**Comment by h-vetinari** [comment]
Ping @isuruf :)

**Comment by h-vetinari** [comment]
@isuruf
WDYT about `linux-64/liblapack-3.9.0-8_netlib_repacked_for_blis.tar.bz2`?

**Comment by h-vetinari** [comment]
@isuruf
I'd like to merge this - I believe I have addressed your concerns, but would like to have confirmation from (or discussion with) you to be sure.

**Comment by isuruf** [comment]
I'm not a fan of this. I believe we can test this by creating the symlinks for blis in the testing script. I'll have a look, but will not be soon.

**Comment by h-vetinari** [comment]
ok, thanks for the response. If you can outline your plan (e.g. symlink what to where), I can give it a shot. In the meantime, I'm placing this PR in draft mode.

**Comment by isuruf** [comment]
Thanks @h-vetinari for all your efforts.

I think what we need to do is in `test_blas.sh` call `build_pkg.sh` for blis four times with `PKG_NAME` set for `blas, cblas, lapack, lapacke`.

**Comment by h-vetinari** [comment]
OK, I'll think about it and might give it a shot. While we're at it, could you briefly comment on [this](https://github.com/conda-forge/blas-feedstock/pull/62/files#r554271759)? Circular dependencies sound bad to me, and should probably be fixed ASAP independently of the blis-stuff.

File: recipe/meta.yaml (hunk 1 above)
Comment (isuruf) [reply recipe/meta.yaml:206]: Yes that's a mistake and should be removed. `blas` should depend on `blas-devel`.

Comment (h-vetinari) [reply recipe/meta.yaml:206]: I thought it was the other way around? Kinda like `blas-devel` > `blas` (in terms of dependencies)

At least that's how I've understood the `-dev[el]` packages in various distros, for example.

**Comment by h-vetinari** [comment]
Obsoleted by #65

**Pull request closed without merging**